### ISSUE TITLE
Get Route Handler Components Asynchronously

### ISSUE THREAD

**Issue by ryanflorence** [rackt/react-router#755]
Right now a route config requires all of the routes to declare their route handle components up front, this is problematic for a couple reasons:

1. Lazy loading
2. Indirect circular deps with the router instance (very problematic in flux, router instance requires components which require actions which require the router instance)

While lazy loading is possible by creating thin handlers like in the `partial-app-loading` example, for cases like Relay's GraphQL queries, we'd still need to be requiring in the child components to get their queries so we're back to having the same problem.

My initial thought is to introduce `getHandlerForRoute` on the router instance. The default implementation would not require anybody to change their existing code but would give us the ability to let the developer provide a component lazily.

```js
var router = Router.create({
  routes: routes,
  getHandlerForRoute (route, state, cb) {
    // this is the default implementation we use if not defined
    cb(route.handler);
  }
});

// if you're using webpack, route handlers could be lazy bundles
<Route handler={require('bundle?lazy./components/FriendsList')}/>
getHandlerForRoute (route, state, cb) {
  route.handler(cb);
}

// if you have some other loading mechanism, you could use strings
<Route handler="FriendsList" />
getHandlerForRoute (route, state, cb) {
  getComponentByString(route.handler, cb);
}
```

I like this a lot, but its simply a strawman to get the conversation going.

**Comment by rpflorence** [comment]
@mjackson @josephsav @cpojer

**Comment by geekyme** [comment]
Looks good and does it require any change of code on server side rendering? Here is a sample that I'm using in yahoo's fluxible architecture https://github.com/geekyme/isomorphic-react/blob/master/server.js#L58-L98

**Comment by rpflorence** [comment]
> does it require any change of code on server side rendering?

All depends on your implementation, we just provide a function and a callback.

**Comment by gaearon** [comment]
This is good. Yes.

What happens before the callback is ready? As you saw in another issue, we ran into a little trouble: if you want transition hooks to work, you can't have a loading UI, because to show the UI, you have to allow the transition.

IMO we need to both allow loading UI *and* let transition hooks work correctly.

**Comment by geekyme** [comment]
> Indirect circular deps with the router instance (very problematic in flux, router instance requires components which require actions which require the router instance)

I didn't know how about that. Please advise. How come actions need the router instance?

**Comment by gaearon** [comment]
>I didn't know how about that. Please advise. How come actions need the router instance?

You might want to `transitionTo`, `goBack` or `replaceWidth` from a Flux action.

**Comment by mjackson** [comment]
> we need to both allow loading UI and let transition hooks work correctly.

Agree. I've done a lot of thinking about this over the weekend. The nice thing about the model we currently have is that we already have a few well-defined points where you get to control 1) how the transition proceeds (the `willTransitionTo` hook) and 2) exactly when you render (the `run` callback). So it doesn't seem like we need any more hooks and/or methods. We probably just need to expose the existing functionality in a way that is more friendly for people who are dynamically loading UI.

To put it another way, we already know how to:

- "Pause" the transition (keep the current UI) until we dynamically load something
- Render some loading UI while we dynamically load something

Which one you do is up to you, according to your needs.

**Comment by gaearon** [comment]
Can route handlers define a static property telling what to show while transition to it is pending?

```js
var ShowNothingUntilImReady = React.createClass({
  statics: {
    willTransitionTo(blabla, callback) {
      setTimeout(callback, 3000);
    },
  }
});

var Spinner = React.createClass({
  render() { return <img src='loading.gif' /> }
});

var ShowSpinnerUntilImReady = React.createClass({
  statics: {
    loadingHandler: Spinner,

    willTransitionTo(blabla, callback) {
      setTimeout(callback, 3000);
    },
  }
});
```

Or should this be declared in route config instead?

```js
<Route name='not-nice' handler={ShowNothingUntilImReady} />
<Route name='nice' handler={ShowSpinnerUntilImReady} loadingHandler={Spinner} />
```

How does this work with nesting?

**Comment by gaearon** [comment]
```js
<Route handler={App}>
  <Route name='nice' handler={ShowSpinnerUntilImReady} />
  <PendingRoute handler={Spinner} />
</Route>
```

Until `ShowSpinnerUntilImReady` calls `willTransitionTo` callback, `App` will render with `Spinner` as `<RouteHandler />`.

If `App` did not have `PendingRoute` inside it, then it transition would freeze it.

This approach probably has problems with more nesting but it's something.

**Comment by tcoopman** [comment]
Should it be the responsibility of the component to chose a spinner or of the router/root component?

So: _component: I'm not done yet, please render this spinner till I'm done._ **vs** _component: I'm not done yet. Router: Ok, while waiting, I will render this spinner_?

Both seem to have advantages.

**Comment by gaearon** [comment]
I think the parent component (or parent's route config node) should decide that. Only parent knows what is appropriate to level inside it at the current level of nesting.

**Comment by plasticine** [comment]
This is really exciting, and definitely solves a really tricky date/lazy use-case that I keep butting into. Awesome.

**Comment by plasticine** [comment]
> Until ShowSpinnerUntilImReady calls willTransitionTo callback, App will get Spinner as <RouteHandler />. If App did not have PendingRoute inside it, then the transition would freeze it (like it does now).

I really like this idea.

This would also allow for managing any initial data load that may also occur after the initial lazy-load of the handler to happen behind the same spinner.

**Comment by ryanflorence** [comment]
@gaearon I know its related, but can we move the discussion around "loading states" to a new thread? We can talk about in the context of waiting on a transition already, I'd like to try to keep this thread dedicated to a solid API for async component loading, and then we can reference this in the other thread.

**Comment by nickdima** [comment]
@ryanflorence your proposal looks good to me. I bumped into this problem just yesterday and I was looking for exactly this implementation.

**Comment by jsdf** [comment]
I think this could be an opportunity to add a lot of flexibility to how matched routes are handled. I'm not even specifically interested for this for the sake of loading routes asynchronously, but I think this same feature could be really useful anywhere you want to be able to dynamically build a handler given the contextual information available in the matched route. 

Doing that with the current API is awkward, but if in the suggested `getHandlerForRoute` the callback were to either take a handler component+props, or an instantiated handler component (which you could still inspect statically, and then cloneWithProps) then route handlers would become much more expressive.

**Comment by pburtchaell** [comment]
👍 This is good. I'm currently building an app that would benefit enormously from this API. 

We should start a new thread to discuss how this API would be connected to transition hooks and the loading component idea. IMO that is a very important part of this.

**Comment by ofersadgat** [comment]
@ryanflorence An alternate idea is to let each handler declare its own routes, then when one is dynamically loaded, it can then load its own routes. I have a rough prototype of this which only requires one small change: https://github.com/rackt/react-router/pull/798 . The way it works is by having a handler add or alter its own routes, then call replaceRoutes in the Router with the root and it will then reload all of the new routes. Not the most efficient implementation, but if you like this path, it can be made more efficient.

**Comment by josephsavona** [comment]
@ryanflorence @mjackson this looks good. It would probably be necessary to be able to abort a call to `getHandlerForRoute`, for example when the user attempts a second transition before the first completes. This would be especially helpful if there's a lot of code for the route and/or you're on a slow connection.

**Comment by globexdesigns** [comment]
I have another use case for async loading which I'm hoping this issue will be able to address in the future. My app's flow is as such:

```
 +------------------------------+
 |   GET request to /internal   |
 +------------------------------+
                |
                V
    +------------------------+         // At this point, the only routes
    |   React.Router.run()   |         // the Router known about are
    +------------------------+         // root (/) and 404.
                |
                V
 +--------------------------------+    // The main app is wrapped in an
 |  willTransition() for AppView  |    // authentication mixin. So we
 +--------------------------------+    // check if the user is logged in.
                |
                V
   +--------------------------+    +----+    +---------------------+
   |  Check if authenticated  |--->| No |--->| redirect to /login  |
   +--------------------------+    +----+    +---------------------+
                |
                V
             +-----+
		     | Yes |
			 +-----+
				|
                V
   +----------------------------+     // Webpack async bundles are only
   |  Dynamically load modules  |     // loaded only if the user is authenticated.
   +----------------------------+     // The bundles contain additional routes.
   				|
                V
 +-------------------------------+    // One of the loaded modules will
 |   Add routes to React-Router  |    // extend the React.Router with the
 +-------------------------------+    // Component for the /internal route
                |
                V
 +-------------------------------+    // Finally, we should load the Component
 | Load Component for /internal  |    // for the /internal route. Currently this
 +-------------------------------+    // fails because there's no way to
                                      // add routes in the middle of a transition
									  // in the previous step.
```

**Comment by gaearon** [comment]
I think new API should solve this with async route configuration.

**Comment by ryanflorence** [comment]
1.0 does all of this and MORE SEND US MONEY!

**Comment by globexdesigns** [comment]
@ryanflorence Cool! Do you have any docs yet for the new async stuff in 1.0?

**Comment by goldensunliu** [comment]
+1

**Comment by Duan112358** [comment]
smell nice
